lib/runner.js

```diff
@@ -303,9 +301,14 @@ class Runner extends Emittery {
 
 	async runTest(task, contextRef) {
 		let hooksAndTestOk = false;
-
 		const hookSuffix = ` for ${task.title}`;
-		if (await this.runHooks(this.tasks.beforeEach, contextRef, hookSuffix)) {
+		let testArgs = task.args || [];
+		// Is it a macro? Then skip its args
+		if (task.implementation && task.implementation.title) {
```

Comment:
Macros are not required to implement `.title`. They're not really distinguishable from regular test implementations.

Reply:
I think I tried to nail down the very last argument of a test (`{initDB: true} in the example`) and only pass it to the hook, since stuff like `'3 * 3', 9` won't be useful. Also document it and explain users that the last argument is always passed to the beforeEach/afterEach hooks.
```
test('providedTitle', macro1, '3 * 3', 9, {initDB: true});
```

But due to inability to really detect when something is a macro I'm puzzled for now. 
And with my implementation args are and array, need to think about @Gwenio response.

Reply:
> I think I tried to nail down the very last argument of a test (`{initDB: true} in the example`) and only pass it to the hook, since stuff like `'3 * 3', 9` won't be useful. Also document it and explain users that the last argument is always passed to the beforeEach/afterEach hooks.
> 
> ```
> test('providedTitle', macro1, '3 * 3', 9, {initDB: true});
> ```
> 
> But due to inability to really detect when something is a macro I'm puzzled for now.
> And with my implementation args are and array, need to think about @Gwenio response.

If both hook args and macro args can both be of variable length then finding where to split them is not possible.
This means to support macros with rest or optional parameters that the hook arg count needs to be fixed.
Having them be fixed not only excludes a rest parameter, but also means the hook arg(s) cannot be optional.
And requiring a hook arg would be an interface breaking change.

To avoid this the hook arg would need to come before the test implementation(s).
Then it could be extracted when getting the optional base title and the implementation(s).
It would also need the type of the arg limited such that it is distinct from the implementation(s) (and the title, if it can come before that too).

The relevant part of the runner for extracting title and implementation(s) is line 60 to 65 of runner.js.
The hook arg should then be added to the task record when it is created on line 129.